spec/features/staff/users_can_view_reports_spec.rb

```diff
@@ -209,6 +209,26 @@ def expect_to_see_grouped_rows_of_reports_for_an_organisation(organisation:, exp
         click_on "Download report as CSV file"
       end
     end
+
+    scenario "they can view all comments made during a reporting period" do
+      report = create(:report)
+
+      activities = create_list(:project_activity, 2, organisation: report.organisation, source_fund_code: report.fund.source_fund_code)
+
+      comments_for_report = [
+        create_list(:comment, 3, activity: activities[0], report: report),
+        create_list(:comment, 1, activity: activities[1], report: report),
+      ].flatten
+
+      page = ReportPage.new(report)
+      page.visit_comments_page
+
+      expect(page).to have_comments_grouped_by_activity(
+        activities: activities,
+        comments: comments_for_report
+      )
+      expect(page).to have_no_edit_buttons_for_comments(comments_for_report)
```

Comment:
I think we could avoid have a negative version of the `has_edit_buttons_for_comments?` matcher if we removed the expectation from the body and use `Enumerable#all?` e.g. 

```ruby
  def has_edit_buttons_for_comments?(comments)
    within ".govuk-table" do
      comments.all? do |comment|
        page.has_link?(href: edit_activity_comment_path(comment.activity, comment))
      end
    end
  end

```

In this way the negative assertion in the feature spec could be made with

`expect(page).not_to have_edit_buttons_for_comments(comments_for_report)`

But I'm not sure it's worth it? What do you think?

Reply:
Oh definitely. I did want to do it that way, but had some issues (I forget what now! 😆). This approach is much nicer.